apiserver/facades/client/cloud/cloud.go

```diff
@@ -1012,6 +1008,23 @@ func (api *CloudAPI) AddCloud(cloudArgs params.AddCloudArgs) error {
 		}
 	}
 
+	if cloudArgs.Cloud.Type != string(provider.K8s_ProviderType) {
+		// All non-k8s cloud need to go through whitelist.
+		controllerInfo, err := api.backend.ControllerInfo()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		controllerCloud, err := api.backend.Cloud(controllerInfo.CloudName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if err := cloud.CurrentWhiteList().Check(controllerCloud.Type, cloudArgs.Cloud.Type); err != nil {
+			if cloudArgs.Force == nil || !*cloudArgs.Force {
+				return errors.Trace(err)
```

Comment:
We need to return a typed error and have the caller (eg CLI) interpret that error and print the desired message relevant to the call context. Like is done for CodeAlreadyExists. The message can be taylored by the CLI to mention the use of --force etc